### ISSUE TITLE
Black Speckles

### ISSUE THREAD

**Issue by william-silversmith** [seung-lab/neuroglancer#71]
Another Storage related bug (in the newest threaded version too).

Sometimes black speckles appear during the upload phase of processing an image. e.g. here's a picture of downsampling on s1_v0.1 using Storage and CloudVolume:

![image](https://cloud.githubusercontent.com/assets/2517065/25862529/6161a870-34b6-11e7-86b7-0fb3d0ca23f6.png)

Some errors from the web inspector:

```
Failed to decode chunk precomputed:volume:https://neuromancer-seung-import.appspot.com/blob/s1_v0.1/segmentation//192_192_30:5,2,1: Error: Raw-format chunk is 1048576 bytes, but 258048 * 4 = 1032192 bytes are expected.
/chunk_worker.bundle.js:81 Failed to decode chunk precomputed:volume:https://neuromancer-seung-import.appspot.com/blob/s1_v0.1/segmentation//192_192_30:5,2,1: Error: Raw-format chunk is 1048576 bytes, but 258048 * 4 = 1032192 bytes are expected.
```   

This indicates that the black spots contain data, but the data may, for example, be truncated.

**Comment by tartavull** [comment]
Do you know if those chunks has the correct content-encoding?

**Comment by william-silversmith** [comment]
Not to self: Checked again, decode failures didn't appear. This might be a black chunk: https://neuromancer-seung-import.appspot.com/blob/s1_v0.1/image/12_12_30/2112-2176_2624-2688_768-832

**Comment by william-silversmith** [comment]
@tartavull I think you're right. I inspected one on gs://neuroglancer/pinky40_v11/watershed_mst_smc_sem5_remap and gzip encoding wasn't set.

**Comment by william-silversmith** [comment]
@tartavull I found another possible cause. Some patches might be set with incorrect permissions. For instance, when I view the base layer (6_6_30) with glance://, I can see it, when I use gs://neuroglancer/ I get a `403 Forbidden` for that tile.

**Comment by william-silversmith** [comment]
I think these are two separate ways that this can happen.

**Comment by william-silversmith** [comment]
I think I know how to fix the content-encoding issue. Right now, content encoding is set via 
```
        blob.upload_from_string(content)
        if compress:
            blob.content_encoding = "gzip"
            blob.patch()
```
This is two HTTP requests. If the first succeeds but the second fails, there will be an update anomaly. The correct way to do it would be to set the Content-Encoding metadata on the same request that uploads the file.

Unfortunately, it looks like the google-cloud-python package is missing this feature. I'll add it in a separate PR.

**Comment by william-silversmith** [comment]
Actually, it might be fixed in the latest google.cloud (we're on your modified version of 0.22). We should probably upgrade to 0.25.

**Comment by tartavull** [comment]
If the second http requests fails the task should be re scheduled until both of them succeeds

**Comment by william-silversmith** [comment]
That's another good point. I'm not sure why that didn't happen. Maybe the exception (if it existed) only killed the thread and didn't propagate to task_exectution. There are so many wonderful bugs here! This should be two PRs if not three.

**Comment by william-silversmith** [comment]
This problem had three causes, one of which is waiting on v0.26 of google-cloud-python:

1) Network issues triggering a zero-fill, with EmptyVolumeException this is a thing of the past.
2) Content-encoding not being set correctly for gzipped files (waiting for v0.26)
3) Not retrying on network failures, there's an upcoming PR for this

**Comment by tartavull** [comment]
Are these issues solved? @william-silversmith

**Comment by william-silversmith** [comment]
These issues are solved. We're at v0.27 of google-cloud-python and we have retry implemented.
